test/compressionworker.ts

```diff
@@ -0,0 +1,222 @@
+import { getMockEnvelope, getMockEvent, MockEventName, observeEvents } from "./utils";
+
+import * as chai from "chai";
+
+type TestWorkerMessageHandler = (message: IWorkerMessage) => void;
+
+const InstrumentationEventName = "Instrumentation";
+let assert = chai.assert;
+
+describe("Compression Worker Tests", () => {
+  let workerMessages: IWorkerMessage[] = [];
+  let messageHandlers: TestWorkerMessageHandler[] = [];
+  let testFailureTimeout: number = null;
+
+  beforeEach(() => {
+    setupFixture([]);
+    workerMessages = [];
+    messageHandlers = [];
+  });
+  afterEach(() => {
+    cleanupFixture();
+    clearTimeout(testFailureTimeout);
+  });
+
+  it("validates that events are batched for upload correctly when total length is below the limit", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let firstMockEventName = "FirstMockEvent";
+    let secondMockEventName = "SecondMockEvent";
+    let firstMockEvent = getMockEvent(firstMockEventName);
+    let secondMockEvent = getMockEvent(secondMockEventName);
+    let addEventMessage = createAddEventMessage(firstMockEvent);
+    let forceUploadMessage = createForceUploadMessage();
+
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    addEventMessage.event = secondMockEvent;
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(forceUploadMessage));
+
+    function messageHandler(message: IWorkerMessage) {
+      assert.equal(message.type, WorkerMessageType.Upload);
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      assert.equal(payload.events.length, 2);
+      assert.equal(payload.events[0].type, firstMockEventName);
+      assert.equal(payload.events[1].type, secondMockEventName);
+      done();
+    }
+  });
+
+  it("validates that events are batched for upload correctly when total length is above the limit", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let firstMockEventName = "FirstMockEvent";
+    let secondMockEventName = "SecondMockEvent";
+    let firstMockEvent = getMockEvent(firstMockEventName);
+    let secondMockEvent = getMockEvent(secondMockEventName);
+    let firstMockData = Array(Math.round(config.batchLimit * (2 / 3))).join("1");
+    let secondMockData = Array(Math.round(config.batchLimit / 2)).join("2");
+    firstMockEvent.state = { data: firstMockData };
+    secondMockEvent.state = { data: secondMockData };
+    let addEventMessage = createAddEventMessage(firstMockEvent);
+    let forceUploadMessage = createForceUploadMessage();
+
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    addEventMessage.event = secondMockEvent;
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(forceUploadMessage));
+    scheduleTestFailureTimeout();
```

Comment:
Does this set up a timeout that causes the test to fail? Since this is a common pattern in these tests can you create a higher level abstraction that makes writing these easier? Lots of tedious boilerplate in these tests. Example:

Tests can be defined as a sequence of input events and output events and a worker. Write a generic test handler for that case.

Reply:
Ok, so I played around with this idea, but it doesn't look like it has much gain. Tests have custom scenarios and can't ALL be placed in the same exact abstraction. I don't think there are many enough tests in this suite to justify the extra work.

---
In reply to: [136876565](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136876565) [](ancestors = 136876565)